src/bartiq/transform.py

```diff
@@ -0,0 +1,151 @@
+                          {
+                              "swap": {"CNOT": 3},
+                              "arbitrary_z": {"T_gates": "3*log2(1/epsilon) + O(log(log(1/epsilon)))"},
+                              ...
+                          }
+        routine: The program to which the resources will be added.
+
+    Returns:
+        Routine: The program with aggregated resources.
+
+    Raises:
+        TypeError: If the input types are not valid bartiq routine.
+    """
+    try:
+        verify_uncompiled_routine(routine, backend=backend)
+    except Exception as e:
+        raise TypeError("Must apply to a valid bartiq routine.") from e
+
+    expanded_aggregation_dict = _expand_aggregation_dict(aggregation_dict)
+    for subroutine in routine.walk():
+        _add_aggregated_resources_to_subroutine(expanded_aggregation_dict, subroutine)
+    return routine
+
+
+def _add_aggregated_resources_to_subroutine(
+    expanded_aggregation_dict: Dict[str, Dict[str, Any]], subroutine: Routine, backend=BACKEND
+) -> Routine:
+    if not hasattr(subroutine, "resources") or not subroutine.resources:
+        return subroutine
+
+    aggregated_resources = copy.copy(subroutine.resources)
+    for resource_name in subroutine.resources:
+        resource_expr = backend.as_expression(subroutine.resources[resource_name].value)
+        if resource_name in expanded_aggregation_dict:
+            mapping = expanded_aggregation_dict[resource_name]
+            for sub_res, multiplier in mapping.items():
+                multiplier_expr = backend.as_expression(multiplier)
+                if sub_res in aggregated_resources:
+                    current_value_expr = backend.as_expression(aggregated_resources[sub_res].value)
+                    aggregated_resources[sub_res].value = str(current_value_expr + multiplier_expr * resource_expr)
+                else:
+                    new_resource = Resource(
+                        name=sub_res,
+                        type=subroutine.resources[resource_name].type,
+                        value=str(multiplier_expr * resource_expr),
+                    )
+                    aggregated_resources[sub_res] = new_resource
+
+            del aggregated_resources[resource_name]
+
+    subroutine.resources = aggregated_resources
+    return subroutine
+
+
+def _expand_aggregation_dict(aggregation_dict: Dict[str, Dict[str, Any]], backend=BACKEND) -> Dict[str, Dict[str, Any]]:
+    """
+    Expand the aggregation dictionary to handle nested resources.
+    Args:
+        aggregation_dict: The input aggregation dictionary.
+    Returns:
+        Dict[str, Dict[str, Any]]: The expanded aggregation dictionary.
+    """
+    if not isinstance(aggregation_dict, dict):
+        raise TypeError("aggregation_dict must be a dictionary.")
+
+    expanded_dict = {}
+    for resource in aggregation_dict.keys():
+        expanded_dict[resource] = _expand_resource(resource, aggregation_dict, set())
+    return expanded_dict
```

Comment:
**Issue:** While correct, the recursive approach taken here can quickly blow up the number of calls. Consider the following, very artificial but still moderately sized, `aggregation_dict`:

```python
{
  "a": {
    "b": 2,
    "c": 3
  },
  "b": {
    "c": 10,
    "d": 4
  },
  "c": {
    "d": 5
  },
  "d": {
    "e": 5
  }
}
```
One would expect that the number of `_expand_resource` calls would be roughly equal to number of resources, meanwhile the following sequence of expansion is observed: "a b c d d c d b c d d c d d". This is not surprising, everytime e.g. "c" is needed it is expanded again, even if it was expanded previously. 

This problem might not sound that bad, but we are parsing expressions on each expansion, and parsing expressions might be expensive, hence I'd rather like we address this before merging.

There are two possible solutions:
- Using memoization.
- Topologically sorting the resources to be expanded, so that the more "primitive" resources get expanded first (I'd prefer this one more).

LMK what you think, I'd be happy to pair on this!

Reply:
@dexter2206 please check the _topological_sort function in the latest commit.